src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -721,26 +738,90 @@ private bool IsAuthenticationError(Exception ex)
             return false;
         }
 
-        private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool success)
+        private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher)
         {
-            TimeSpan cacheExpirationTime;
+            TimeSpan cacheExpirationTime = changeWatcher.CacheExpirationInterval;
+            changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
+        }
 
-            if (success)
-            {
-                changeWatcher.RefreshAttempts = 0;
-                cacheExpirationTime = changeWatcher.CacheExpirationInterval;
-            }
-            else
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(IEnumerable<ConfigurationClient> clients, Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = clients.GetEnumerator();
+
+            clientEnumerator.MoveNext();
+
+            ConfigurationClient currentClient;
+
+            while (true)
             {
-                if (changeWatcher.RefreshAttempts < int.MaxValue)
+                bool success = false;
+                cancellationToken.ThrowIfCancellationRequested();
+                currentClient = clientEnumerator.Current;
+
+                try
+                {
+                    T result = await funcToExecute(currentClient).ConfigureAwait(false);
+                    success = true;
+
+                    return result;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!IsFailOverable(ae) || !clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException rfe)
+                {
+                    if (!IsFailOverable(rfe) || !clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+                finally
                 {
-                    changeWatcher.RefreshAttempts++;
+                    _configClientManager.UpdateClientStatus(currentClient, success);

```

Comment:
For not failoverable errors, for example 401 response, I'm wondering if we should put all available clients to backoff state since we don't extend watcher on failure anymore and we rely on client backoff to not go crazy retrying.

Reply:
I agree that we should backoff on 401 (exponential backoff like other error codes).

> I'm wondering if we should put all available clients to backoff state

Is it guaranteed that if user is not authorized for one replica, they are not authorized for other replicas as well?

Also, the load on the server for retries will be ~5 requests per 10 minutes per replica => ~30 requests per hour per replica. 
In the grand scheme of things, ~30 requests/hour doesn't seem like too much load on the server, so I don't know if we want to backoff all available clients on a 401.

Reply:
When a request is not failoverable, it means we don't think the request can succeed with another replica, right? If we don't think the request will have better luck with another replica, why do we bother to let it try? In the past, we used watcher expiration time to control whether (and when) we should try again if something failed. You changed it to use client back-off for the same functionality now. But watchers are global. They are independent of individual clients, so shouldn't we put all clients in the back-off state to match the behavior?

There is a downside to the current approach. In the past, we tracked the watchers' expiration time individually. When one watcher is extended to retry later, we might still work for another watcher. In the new/current approach, we put the whole client in a backoff state when one watcher failed, which effectively extends all watchers. This is not necessarily wrong because requests are related in some ways, but I want to make sure we realize the change.

Reply:
> When a request is not failoverable, it means we don't think the request can succeed with another replica, right? If we don't think the request will have better luck with another replica, why do we bother to let it try?

I think we should backoff all clients only if the exception is not tied to the specific watcher.
meaning, we can backoff all clients for a 401 exception, but we can't backoff all clients for a 404 exception.

Looking at the HTTP response codes, I think it's only 401 and 407 that qualify to backoff all clients.

Reply:
On another thought, it seems we can backoff on exceptions like 404 as well because watchers are related as you said.

Reply:
I think 404 is considered a success...

It goes to your definition of "failoverable". If you don't think a request is failoverable, why do you want to try again with another client? If you want to try again with another client, why don't you fail it over in the first place? You see the contradiction here?

Reply:
Based on our offline discussions, updating the logic to backoff all clients for non-failoverable errors.